### PR TITLE
feat: License V3 

### PR DESCRIPTION
**Pull request by lmauromb**
## Proposed changes (including videos or screenshots)

- new `package/jwt` with `sign` and `veriy` functions to create a JWT given a JS object.
  - uses [jose](https://github.com/panva/jose)
  -  by default uses algorithm RS256 for the public/private keys.
- new License package, `ee/package/license`
  - new definition V3 and previous License definition will be renamed  as LicenseV2
  - package handles both definitions
  - previous license references were updated to use this new package

## Issue(s)

https://rocketchat.atlassian.net/browse/NBJ-292

### CODE REVIEW

File: apps/meteor/client/views/hooks/useUpgradeTabParams.ts [hunk 1]
```diff
@@ -16,9 +17,14 @@ export const useUpgradeTabParams = (): { tabType: UpgradeTabVariant | false; tri
...
-	const isTrial = licensesData?.licenses?.every(({ meta }) => meta?.trial) ?? false;
-	const trialEndDate = trialLicense?.meta ? format(new Date(trialLicense.meta.trialEnd), 'yyyy-MM-dd') : undefined;
+	const licenses = (licensesData?.licenses || []) as (Partial<ILicenseV2 & ILicenseV3> & { modules: string[] })[];
+
+	const trialLicense = licenses.find(({ meta, information }) => information?.trial ?? meta?.trial);
+	const isTrial = Boolean(trialLicense);
+	const trialEndDate =
+		trialLicense?.meta?.trialEnd || trialLicense?.cloudMeta?.trialEnd
+			? format(new Date(trialLicense.meta?.trialEnd ?? trialLicense.cloudMeta?.trialEnd), 'yyyy-MM-dd')
+			: undefined;
```
Comment (hugocostadev) [inline apps/meteor/client/views/hooks/useUpgradeTabParams.ts:27]: @ggazzo shouldn't be the `visualExpiration` attribute, like we discussed?

File: ee/packages/license/src/license.ts [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+	} catch (e) {
+		logger.error('Invalid license');
+		if (process.env.LICENSE_DEBUG && process.env.LICENSE_DEBUG !== 'false') {
+			logger.error({ msg: 'Invalid raw license', encryptedLicense, e });
+		}
+		return false;
+	}
+};
+
+export const isEnterprise = () => Boolean(license && valid);
```
Comment (hugocostadev) [inline ee/packages/license/src/license.ts:145]: I suggest rename this method to `hasValidLicense`

File: ee/packages/license/src/validation/getCurrentValueForLicenseLimit.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+const dataCounters = new Map<LicenseLimitKind, (context?: LimitContext<LicenseLimitKind>) => Promise<number>>();
+
+export const setLicenseLimitCounter = <T extends LicenseLimitKind>(limitKey: T, fn: (context?: LimitContext<T>) => Promise<number>) => {
+	dataCounters.set(limitKey, fn as (context?: LimitContext<LicenseLimitKind>) => Promise<number>);
+};
+
+setLicenseLimitCounter('activeUsers', () => Users.getActiveLocalUserCount());
+setLicenseLimitCounter('guestUsers', () => Users.getActiveLocalGuestCount());
+setLicenseLimitCounter('roomsPerGuest', async (context) => (context?.userId ? Subscriptions.countByUserId(context.userId) : 0));
```
Comment (ggazzo) [inline ee/packages/license/src/validation/getCurrentValueForLicenseLimit.ts:17]: promote to application layer

File: ee/packages/license/src/showLicense.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+	}
+
+	const {
+		validation: { serverUrls, validPeriods },
+		limits,
+	} = license;
+
+	const modules = getModules();
+
+	console.log('---- License enabled ----');
```
[the remark this answers is not in the document]
Comment (lmauromb) [reply ee/packages/license/src/showLicense.ts:21]: I think we need to use console.log in order to prompt the messages when starting the application without increasing the log level

cc: @pierre-lehnen-rc @ggazzo

File: .changeset/twelve-files-deny.md [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+'@rocket.chat/core-typings': minor
+'@rocket.chat/rest-typings': minor
+'@rocket.chat/ddp-streamer': minor
+'@rocket.chat/queue-worker': minor
+'@rocket.chat/presence': minor
+'@rocket.chat/meteor': minor
+---
+
+feat new `package/jwt` with `sign` and `veriy` functions to create a JWT given a JS object. By default uses algorithm RS256 for the public/private keys.
+feat new `ee/package/license`. Contains new V3 definition and previous License definition renamed as LicenseV2. Package handles both versions.
```
Comment (debdutdeb) [inline .changeset/twelve-files-deny.md:21]: i think this should be just "New license v3 type" or something

File: ee/packages/license/src/showLicense.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ggazzo) [reply ee/packages/license/src/showLicense.ts:21]: Exact, to keep the same behavior, if we use logger this is going to be filtered

File: ee/packages/license/src/validation/getCurrentValueForLicenseLimit.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+		return dataCounters.get(limitKey)?.(context as LimitContext<LicenseLimitKind> | undefined) ?? 0;
+	}
+
+	logger.error({ msg: 'Unable to validate license limit due to missing data counter.', limitKey });
+
+	return 0;
+};
+
+export const hasAllDataCounters = () =>
+	(['activeUsers', 'guestUsers', 'roomsPerGuest', 'privateApps', 'marketplaceApps', 'monthlyActiveContacts'] as LicenseLimitKind[]).every(
```
[the remark this answers is not in the document]
Comment (ggazzo) [reply ee/packages/license/src/validation/getCurrentValueForLicenseLimit.ts]: I'm not sure what you expect to do with this array

[the remark this answers is not in the document]
Comment (ggazzo) [reply ee/packages/license/src/validation/getCurrentValueForLicenseLimit.ts]: use `shouldPreventAction` or the internal methods there is no reason to use or to have access to the array

File: apps/meteor/ee/app/authorization/server/validateUserRoles.js [hunk 7]
```diff
@@ -1,29 +1,42 @@
...
+
+	if (currentUserData?.type === 'app') {
 		return;
 	}
 
-	if (userData.roles.length >= 2) {
-		throw new Meteor.Error('error-guests-cant-have-other-roles', "Guest users can't receive any other role", {
-			method: 'insertOrUpdateUser',
-			field: 'Assign_role',
-		});
```
[the remark this answers is not in the document]
Comment (pierre-lehnen-rc) [reply apps/meteor/ee/app/authorization/server/validateUserRoles.js:19]: the "Guest users can't receive any other role" error was just removed, as this was the only place that validated this; It was never a product decision to do this but if it is, then it needs to be implemented properly anyway.

**Review by ggazzo — APPROVED** [review]

**Pull request merged**